vendor/github.com/robert-milan/go-object-interning/object_intern.go

```diff
@@ -0,0 +1,484 @@
+		return 0, err
+	}
+
+	// remove 4 trailing bytes for reference count
+	return *(*uint32)(unsafe.Pointer(objAddr + uintptr(len(compObj)-4))), nil
+
+}
+
+// ObjBytes returns a []byte and nil on success.
+// On failure it returns nil and an error.
+//
+// If compression is used, this method does not use the interned data to create a []byte,
+// instead it allocates a new []byte.
+//
+// If compression is turned off, this will return a []byte slice with the backing array
+// set to the interned data.
+func (oi *ObjectIntern) ObjBytes(objAddr uintptr) ([]byte, error) {
+	oi.RLock()
+	defer oi.RUnlock()
+
+	b, err := oi.Store.Get(objAddr)
+	if err != nil {
+		return nil, err
+	}
+	// remove 4 trailing bytes for reference count and decompress
+	objDecomp, err := oi.decompress(b[:len(b)-4])
+	if err != nil {
+		return nil, err
+	}
+	return objDecomp, nil
+}
+
+// ObjString returns a string and nil on success.
+// On failure it returns an empty string and an error.
+//
+// This method does not use the interned data to create a string,
+// instead it allocates a new string.
+func (oi *ObjectIntern) ObjString(objAddr uintptr) (string, error) {
+	oi.RLock()
+	defer oi.RUnlock()
+
+	b, err := oi.Store.Get(objAddr)
+	if err != nil {
+		return "", err
+	}
+	// remove 4 trailing bytes for reference count and decompress
+	objDecomp, err := oi.decompress(b[:len(b)-4])
+	if err != nil {
+		return "", err
+	}
+	return string(objDecomp), nil
+}
+
+// SetString takes an object adress and a string header.
+// Upon success it sets the string to point at objAddr with the proper length set.
+// On failure it takes no action and returns false.
+func (oi *ObjectIntern) SetString(objAddr uintptr, szHdr *reflect.StringHeader) bool {
+	oi.RLock()
+	defer oi.RUnlock()
+
+	b, err := oi.Store.Get(objAddr)
+	if err != nil {
+		return false
+	}
+
+	objDecomp, err := oi.decompress(b[:len(b)-4])
+	if err != nil {
+		return false
+	}
+
+	szHdr.Data = objAddr
+	szHdr.Len = len(objDecomp)
+	return true
+}
+
+// SetStringNoCprsn takes an object adress and a string header, it assumes that compression is turned off.
+// Upon success it sets the string to point at objAddr with the proper length set.
+// On failure it takes no action and returns false.
+func (oi *ObjectIntern) SetStringNoCprsn(objAddr uintptr, szHdr *reflect.StringHeader) bool {
+	oi.RLock()
+	defer oi.RUnlock()
+
+	b, err := oi.Store.Get(objAddr)
+	if err != nil {
+		return false
+	}
+
+	szHdr.Data = objAddr
+	szHdr.Len = len(b) - 4
+	return true
+}
+
+// LenNoCprsn takes a slice of object addresses, it assumes that compression is turned off.
+// Upon success it returns a slice of the lengths of all of the interned objects - the 4 trailing bytes for reference count, and true.
+// The returned slice indexes should match the indexes of the slice of uintptrs.
+// On failure it returns a possibly partial slice of the lengths, and false.
+func (oi *ObjectIntern) LenNoCprsn(ptrs []uintptr) (retLn []int, all bool) {
```

Comment:
`NoCprsn` seems a bit convoluted as a term. I would suggest `Raw` but that's too vague.
maybe the functions should be named after extra features when appropriate?
E.g. `Len` would assume not compressed, and `LenCompressed` would be for compressed?
maybe we can do the same for `NoRefCnt` -> assume no refcount by default, and if there is one, add `RefCount` to the func name?
note that i'm not really familiar with this code base yet so maybe my suggestion doesn't make sense

Reply:
I agree that the naming conventions are not great. I'm not sure exactly how to change them yet. During the initial design I was operating under the assumption that everything would be compressed, however it turns out that it makes more sense to use uncompressed data. So, originally everything was assumed to be compressed and all retrieval methods increased the reference count. I think the opposite is true now, so your suggestions do make sense to make `NoCprsn` the norm.

Reply:
Ping me when you're ready to do a hangout about this. I want to understand this better.
I also wonder if it makes sense to mix compressed and uncompressed strings in the same go-object-interning instance (or perhaps better to create several instances then)